2018/python/file_generator.py

```diff
@@ -487,7 +487,115 @@ def edit_report(self):
 			filepath=edit_report_path))
 
 
+	def create_custom_row(self, dictionary, clean_filepath, clean_filename):
 
+		"""
+		Creates a custom clean LAR row by passing in a dictionary of columns 
+		and new values to modify all the rows of an 
+		existing clean file, filters the modified file for clean rows, 
+		and then pulls the first row from the file. 
+		Pulls rows from the clean file last generated. Suggestion that 
+		the file pulled should be 1000 original rows or greater, to ensure
+		that modified clean rows can be found. 
+		"""
+		#Creates a TS and LAR dataframe from the clean filepath and name
+		#specified. 
+		ts_df, lar_df = utils.read_data_file(
+					path=clean_filepath,
+					data_file=clean_filename)
+
+		#Changes each column (key in the dictionary) to the new value in 
+		# the dictionary.
+		for key, value in dictionary.items():
+			lar_df[key] = value
+
+		checker = rules_engine(lar_schema=self.lar_schema_df, 
+					ts_schema=self.ts_schema_df, 
+					crosswalk_data=self.crosswalk_data)
+
+		#Produces a report as to which syntax or validity
+		#edits have passed or failed based on logic in the rules_engine.
+		#Loads the TS and LAR dataframes into the checker object. 
+
+		checker.load_data_frames(ts_df, lar_df)
+
+		for func in dir(checker):
+			if func[:1] in ("s", "v") and func[1:4].isdigit()==True:
+				getattr(checker, func)()
+
+	
+		#Produces a report as to which syntax or validity
+		#edits have passed or failed based on logic in the rules_engine.
+		for func in dir(checker):
+			if func[:1] in ("s", "v") and func[1:4].isdigit()==True:
+				getattr(checker, func)()
+		
+		#Creates a results dataframe and keeps the results that 
+		#have failed. 
+		res_df = pd.DataFrame(checker.results)
+		new_df = res_df[(res_df['status']=='failed')]
+
+		# The function ignores TS edits and drops results related
+		# to edit fails from the TS.  
+		new_df = new_df[new_df['row_ids'] != 'TS']
+
+		if len(new_df) == 0:
+			#If there are no syntax or validity edits
+			#the data is written to a new directory for quality 
+			#test files that pass syntax and validity edits. 
+
+			utils.write_file(path=self.filepaths['quality_pass_s_v_filepath'].format(
+				bank_name=self.data_map['name']['value']), 
+				ts_input=ts_df, 
+				lar_input=lar_df, 
+				name=quality_filename)
+
+		#The case if there are rows that failed syntax or validity edits.
+		
+		else: 
+			#Creates an empty list for storing row numbers
+			#where edits have failed. 
+			uli_list = []
```

Comment:
Are these ULIs available in res_df where status = failed?

If so, the code to extract them is not needed.